### DISCUSSION TITLE
Can I reconstruct an image from the graph?

### DISCUSSION THREAD

**Discussion by yc-cui** [pyg-team/pytorch_geometric#4254] [Q&A]
@[Somedaywilldo](https://github.com/Somedaywilldo)  visualizes the mnist_superpixel: [https//github.com/pyg-team/pytorch_geometric/issues/320#issuecomment-828382760](https://github.com/pyg-team/pytorch_geometric/issues/320#issuecomment-828382760)

My questions is whether the mnist_superpixel can be reconstructed to the origial RGB image?

**Comment by rusty1s** [comment]
Thanks for the issue. The superpixel graph is a reduced/sparse representation of the original RGB image (the image is reduced from 28x28 pixel to 75 pixels), so there is no way to reconstruct it in a lossless fashion.

For MNIST, you could create a Voronoi diagram from the given point cloud, and color each cell with the color of the original node. This should give you a pretty good visualization of the original MNIST image.

**Reply by yc-cui** [reply]
Thanks for your quick reply. I like PyG, it's quite a fantastic and useful library to me. But Is there any way to use the Gcn layer just like Cnn? I mean, can I just input an image to Gcn layer and then it outputs some features? I want to replace the Cnn layer ，using some layer in PyG.

**Reply by rusty1s** [reply]
For this, you would need to convert your image to a graph first. You can have a look at `torch_geometric.utils.grid` for this. Other than that, this is totally possible.

However, if you have an image there is no good reason to use a GNN over a CNN to process it, especially in terms of efficiency. CNNs are a dedicated class of architectures especially designed for image processing, while GNNs denote a much broader class of architectures.

**Reply by yc-cui** [reply]
Thank you, @rusty1s.  I will have a try. The reason why I want to do that is that in this paper: [https://arxiv.org/pdf/2010.13914.pdf](https://arxiv.org/pdf/2010.13914.pdf), they process the incomplete images using GNN. However, they didn't release the code. So I want to implement it using PyG. BTW, is there any examples about `torch_geometric.utils.grid`?

**Reply by rusty1s** [reply]
We do not have an official example for this but I once created these two gists:
* https://gist.github.com/rusty1s/74b1efe973482e070c50f650ece82b70
* https://gist.github.com/rusty1s/866ca8b830f22aa34923a0f8164f6e64

Hopefully, they are helpful to you.

**Reply by yc-cui** [reply]
Many thanks！

**Reply by yc-cui** [reply]
> We do not have an official example for this but I once created these two gists:
> 
> * https://gist.github.com/rusty1s/74b1efe973482e070c50f650ece82b70
> * https://gist.github.com/rusty1s/866ca8b830f22aa34923a0f8164f6e64
> 
> Hopefully, they are helpful to you.

Hello, @rusty1s . Sorry to bother you again. I have seen these two gists. May I ask how could I combine Dataloaders in PyTorch and in PyG?  The batch from Dataloader in PyG is different from Dataloader in PyTorch. In PyTorch, it's like `[B, C, H, W]`. I have implemented a part of my network (which deals with images using CNN) using PyTorch, and I want to insert some GNN layers into the network. However, the input of the GNN layers is the output of the CNN which is `[B, C, H, W]`. How could I deal with it in this situation?

**Reply by rusty1s** [reply]
For a GNN layer in PyG, it means that `B`, `C` and `H` denote the nodes in your graph (for which you want to share parameters). As such, if `x` is given by `[B, C, H, W]` you need to convert it to `[B * H * W, C]` via `x.permute(0, 3, 1, 2).view(-1, num_features)`.

**Reply by yc-cui** [reply]
Oh! I see. Many thanks again :)
